packages/node-core/src/api.service.ts

```diff
@@ -2,43 +2,57 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import {NetworkMetadataPayload} from './events';
-import {ISubqueryProject} from './indexer';
+import {ApiConnectionError, ApiErrorType, ConnectionPoolService, ISubqueryProject} from './indexer';
 import {getLogger} from './logger';
 
 const logger = getLogger('api');
 
 const MAX_RECONNECT_ATTEMPTS = 5;
 
-type FetchFunction<T> = (batch: number[]) => Promise<T[]>;
-type FetchFunctionProvider<T> = () => FetchFunction<T>;
-
-export abstract class ApiService<P extends ISubqueryProject = ISubqueryProject, A = any, B = any> {
-  constructor(protected project: P) {}
+export interface IApi<A = any, SA = any, B = any> {
+  fetchBlocks(heights: number[], ...args: any): Promise<B[]>;
+  safeApi(height: number): SA;
+  unsafeApi: A;
+  handleError?(error: Error): ApiConnectionError;
+  apiConnect?(): Promise<void>;
+  apiDisconnect?(): Promise<void>;
```

Comment:
I don't think optional properties on an interface is a good idea, either we implement them or we have a different interface that extends this. 

In this case, the exposed IApi that is consumed outside of ApiService/ConnectionPool such as fetch and project service shouldn't be aware of these but the connection specific (apiPromise.connection) these should be required properties that implement an extended interface